File: leetcode.com/2816_double_a_number_represented_as_a_linked_list/solution.c
int solve(struct ListNode *node)
{
    if (!node)
        return 0;

    int r = solve(node->next);
    int n = node->val*2 + r;

    node->val = n % 10;
    return n / 10;
}

struct ListNode *doubleIt(struct ListNode *head)
{
    int r = solve(head);
    if (r == 0)
        return head;

    struct ListNode *new_head = malloc(sizeof(struct ListNode));
    new_head->val = r;
    new_head->next = head;
    return new_head;
}

File: leetcode.com/2816_double_a_number_represented_as_a_linked_list/solution.py
class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        digits = []
        node = head
        while node:
            digits.append(str(node.val))
            node = node.next

        sys.set_int_max_str_digits(20000)
        number = str(int(''.join(digits))*2)

        node = head = ListNode()
        for d in number:
            node.next = ListNode(d)
            node = node.next

        return head.next
